Add CSM model

eustlb opened the pull request:
# What does this PR do?

Adds CSM (audio)-text-to-speech model! 
[Original code](https://github.com/SesameAILabs/csm)
[Hub model weights](https://huggingface.co/sesame/csm-1b)
[Converted weights](https://huggingface.co/eustlb/csm-1b)

eustlb committed:
draft structure

eustlb committed:
depth decoder with forward pre hook

eustlb committed:
full model forward draft

eustlb committed:
draft update

eustlb committed:
depth decoder update

eustlb committed:
ConversationalSpeechModelForCausalLM udpates

eustlb committed:
add generate

eustlb committed:
max length criteria small fix

eustlb committed:
udpate

eustlb committed:
updates

eustlb committed:
generation update

eustlb committed:
update in loss compute

eustlb committed:
conversion script

eustlb committed:
update for correct input embeddings

eustlb committed:
handle interleaved rope

eustlb committed:
update

eustlb committed:
update

eustlb committed:
update

eustlb committed:
support compile

eustlb committed:
update training

eustlb committed:
add doc

eustlb committed:
update doc

eustlb committed:
correct inits

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_36719). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

SeungyounShin commented:
Do you mind if I contribute to CSMForConditionalGeneration? It looks like it's currently empty.

eustlb commented:
Hey @SeungyounShin, currently, conditional generation is handled through the `ForCausalLM` class. Actually, for CSM it makes no difference if the model is generated from context (text + audio) or from text only. I decided for go with the `ForCausalLM` naming because of that, usually, the `ForConditionalGeneration` class is for encoder-decoder like architectures, but this is up to debate with the core maintainers as soon as the PR is ready for review 😊 
thanks a lot for offering your help 🤗

eustlb committed:
Merge branch 'main' into add-csm

SeungyounShin commented on src/transformers/models/conversational_speech_model/modeling_conversational_speech_model.py:
```diff
@@ -0,0 +1,1907 @@
+
+        # as itself generate does not handle input_ids with a depth dimension (here we have [batch_size, seq_length, num_codebooks])
+        # we circumvent this by providing the inputs_embeds directly that allows use to fall back to an handled shape [batch_size, seq_length, hidden_size]
+        inputs_embeds = self.backbone_model.get_input_embeddings()(input_ids)
```
As far as I know, `input_embeds` should have the shape `[batch_size, seq_length, codebook_dim, hidden_size]`. Why does it currently have the shape `[batch_size, seq_length, hidden_size]` instead?

eustlb replied:
the backbone model follows LLama architecture. It expects `input_embeds` of shape `(BS, seq_len, hidden_size)`.
Nevertheless, CSM `input_ids` are in the shape `(BS, seq_len, num_codebooks + 1)`. Were we to simply embed those, we would get something like `(BS, seq_len, num_codebooks + 1, hidden_size)`. To go back to the expected `input_embeds`, CSM simply sum along the 3rd dimension.

BTW I tested my implementation (internally, tests soon to be added) both on logits and generated `input_ids` 😊

SeungyounShin replied:
Thanks for elaborating! But still...

can you test : 
```python
from transformers import AutoModelForCausalLM, AutoTokenizer
import torch
from tokenizers.processors import TemplateProcessing


PATH = './eustlb/csm-1b'


def load_llama3_tokenizer():
    """
    https://github.com/huggingface/transformers/issues/22794#issuecomment-2092623992
    """
    tokenizer_name = "meta-llama/Llama-3.2-1B"
    tokenizer = AutoTokenizer.from_pretrained(tokenizer_name)
    bos = tokenizer.bos_token
    eos = tokenizer.eos_token
    tokenizer._tokenizer.post_processor = TemplateProcessing(
        single=f"{bos}:0 $A:0 {eos}:0",
        pair=f"{bos}:0 $A:0 {eos}:0 {bos}:1 $B:1 {eos}:1",
        special_tokens=[(f"{bos}", tokenizer.bos_token_id), (f"{eos}", tokenizer.eos_token_id)],
    )

    return tokenizer

model = AutoModelForCausalLM.from_pretrained(PATH)
tokenizer = load_llama3_tokenizer()

# (1,L,33)
input_ids = torch.load("./inputs/curr_tokens.pt").cpu()

output = model.generate(input_ids=input_ids, max_new_tokens=25, do_sample=True)

# decode text
print(output.shape)
```

[curr_tokens.pt.zip](https://github.com/user-attachments/files/19645092/curr_tokens.pt.zip)


I always got <reserved_token_1> no !(which is 0 for tokenizer)

eustlb replied:
Indeed ! That is simply because you used `input_ids` directly from Sesame implementation. In the `input_ids` you've provided, codebook tokens for text frames (your input is text frames only) are padded with 0 (which is the original Sesame implementation). In the transformers implementation, we do not use 0 has padding token for codebook has it can actually be predicted but rather something that is not in the codec vobab, i.e. `2050`.  This allows correct construction of the `input_embeds` as explained above. 

All that is going to be documented (and handled by the processor), so just show a tiny bit of patience 🤗

SeungyounShin replied:
Thanks for clarifying, @eustlb! I misunderstood that detail about the padding token difference between Sesame and your transformers implementation. Looking forward to the documentation and updated processor! Appreciate your patience and thorough explanations 🙌

eustlb committed:
ConversationalSpeechModel -> Csm

eustlb committed:
conf update

eustlb committed:
name update

eustlb committed:
tests CsmForCausalLMTest

eustlb committed:
convert use cached_file

eustlb committed:
conf + modeling updates

eustlb committed:
generate utils handle third dim shape

eustlb committed:
integration test

eustlb committed:
modeling + conf updates

eustlb committed:
common test handle more than 2 dims

eustlb committed:
add nested audio list utils

eustlb committed:
processing handle nested audio list

eustlb committed:
csm processing draft

eustlb committed:
mimi util

eustlb committed:
init updates

eustlb committed:
modular update

eustlb committed:
convert modular

eustlb committed:
processing update

eustlb committed:
csm tests update

eustlb committed:
generate tests handle third dim

eustlb committed:
generate utils handle third dim

eustlb committed:
propagate _get_initial_cache_position update

eustlb committed:
tied_weight_keys update + convert correctly

eustlb committed:
fix inputs_embeds

eustlb committed:
revert audio nested list

thomasgauthier commented:
@eustlb I wanted to open a PR for my [own transformers implementation of CSM](https://github.com/thomasgauthier/csm-hf) but I realize now that would be redundant given this PR. I still care about seeing CSM support in transformers so let me know if I can help in any way!

eustlb committed:
batch inference update + return audio

eustlb committed:
audio_utils update

eustlb committed:
processor update

eustlb committed:
some more integration tests

eustlb commented:
@thomasgauthier Nice work!! Cool to see you also went with a chat template for context-speech interleaving 😊
This implementation is a bit more tailored to our API. I’ll keep you posted if I spot anything where your help could be useful, thanks! 🤗

eustlb committed:
remove old test

eustlb committed:
porcessing output labels

eustlb committed:
improve

eustlb committed:
fix

eustlb committed:
update rope values with equivalent ones

eustlb committed:
conversion update

eustlb committed:
udpate tests

eustlb committed:
handle depth decoder generation config

eustlb committed:
remove default eos_token_id

eustlb committed:
make style

eustlb committed:
revert modeling_mimi

eustlb committed:
add default generation_config

eustlb committed:
remove sdpa since handled by default

eustlb committed:
make

eustlb committed:
fix conflict

eustlb committed:
fix conflicts

eustlb committed:
Merge branch 'main' into add-csm

eustlb committed:
correct naming

eustlb committed:
correct imports

eustlb committed:
make

eustlb committed:
causal -> conditional naming

eustlb committed:
causal -> conditional naming

eustlb committed:
Merge branch 'main' into add-csm

eustlb committed:
auto update

eustlb committed:
make

eustlb committed:
make

eustlb committed:
Merge branch 'main' into add-csm

eustlb committed:
add doc

eustlb committed:
Merge branch 'main' into add-csm

eustlb committed:
test update

eustlb committed:
fix weight init

eustlb committed:
audio tokens offsets as buffer

eustlb committed:
4d mask in conditional class

eustlb committed:
make

eustlb committed:
Merge branch 'main' into add-csm

eustlb committed:
Merge branch 'main' into add-csm

eustlb committed:
Merge branch 'main' into add-csm

eustlb committed:
doc update

eustlb committed:
fix causal mask

eustlb committed:
fix causal mask

eustlb committed:
doc update

eustlb committed:
doc update

eustlb committed:
Merge branch 'main' into add-csm

eustlb commented:
Hey @patcon, thanks for the feedback! Unfortunately, we do not go with this naming convention. Nevertheless such search should work as soon as model doc will be update with mentions to Sesame 😊

eustlb committed:
add processor doc

eustlb committed:
update doc

eustlb committed:
fix 4d causal mask

eustlb committed:
update make_list_of_audio

eustlb commented on src/transformers/models/csm/modeling_csm.py:
````diff
@@ -0,0 +1,2158 @@
+        >>> sf.write("output.wav", audio_values[0].cpu().numpy(), 24000)
+        ```
+        """
+        # TODO: ensure the user is not requesting an unsupported generation mode (!= greedy/ sampling)
+        # TODO: ensure not using unsupported logits processors
````
cc @gante, **is there a recommended way of doing this?** I can override `_get_logits_processor` (as I did for `_get_stopping_criteria `), but it is a bit trickier for the generation mode, as it is retrieved from a GenerationConfig method.

gante replied:
uhmmm good question 🤔 

If the list of supported processors is small, perhaps we can whitelist the allowed flags (and throw warnings on the others?). 

Otherwise, overwriting `_get_logits_processor` is probably the best option.

__________________

Out of curiosity, what is the cause of the limited logit processor support in this model?

eustlb replied:
I left that in the beginning but after checking their should not be unsupported logits processor indeed :) 
For the generation mode, I ended up monkey patching get_generation_mode in overridden `_prepare_generation_config`:
```python
def _prepare_generation_config(...):
        ...
        # Monkey patch the get_generation_mode method to support CSM model
        original_get_generation_mode = generation_config.get_generation_mode

        def patched_get_generation_mode(assistant_model=None):
            generation_mode = original_get_generation_mode(assistant_model)
            if generation_mode not in [GenerationMode.GREEDY_SEARCH, GenerationMode.SAMPLE]:
                raise ValueError(
                    f"Generation mode {generation_mode} is not supported for CSM model. Please set generation parameters to use greedy or sampling generation."
                )

            return generation_mode

        generation_config.get_generation_mode = patched_get_generation_mode
```

eustlb commented on src/transformers/generation/utils.py:
```diff
@@ -1701,7 +1703,7 @@ def _prepare_generation_config(
         return generation_config, model_kwargs
 
-    def _get_initial_cache_position(self, input_ids, model_kwargs):
+    def _get_initial_cache_position(self, seq_length, device, model_kwargs):
```
cc @gante

eustlb reviewed (commented):
Ready for review @Cyrilvallez 🤗

Also pinging @gante here:
1. I had to do a few little changes on generate's code that ensures we're able to handle `input_ids` with a third dimension. Basically it replacements of `batch_size, seq_length = input_ids.shape` for `input_ids.shape[:2]` and `input_ids.shape[-1]` for `input_ids.shape[1]`
2. `_get_initial_cache_position` →  `cache_position` initialization directly from `seq_length` and `device`

eustlb committed:
Merge branch 'main' into add-csm

ArthurZucker commented:
Before I review just want to test copilot perfs

gante commented:
@eustlb the `generate`-side changes LGTM 👍  I've replied to your other question directly to your comment in `models/csm/modeling_csm.py`

ArthurZucker commented on src/transformers/models/csm/modular_csm.py:
```diff
@@ -0,0 +1,1517 @@
+        is_first_generation_step = cache_position[0] == 0
+        if is_first_generation_step:
+            # adds place holder in position 0 that will be replaced by the backbone_last_hidden_state
+            model_inputs["input_ids"] = nn.functional.pad(model_inputs["input_ids"], (1, 0), value=0)
```
the tokenizer should just add the token it is more explicit

ArthurZucker replied:
this is weird

eustlb replied:
Agree but there is no tokenizer here, since at each step:
1. the backbone model generates the first codebook token id
2. this token serves as input_id for the depth decoder, and the last backbone hidden state in prepended to the `inputs_embeds`

Nevertheless, I agree it's clearer to directly concatenate this placeholder token before calling generate on the depth decoder. Changing accordingly 😊

ArthurZucker commented on src/transformers/models/csm/modular_csm.py:
```diff
@@ -0,0 +1,1517 @@
+            )
+
+    @can_return_tuple
+    @deprecate_kwarg("num_logits_to_keep", version="4.50", new_name="logits_to_keep")
```
not needed remove it directly

ArthurZucker commented on src/transformers/models/csm/modular_csm.py:
```diff
@@ -0,0 +1,1517 @@
+                model_kwargs["encoder_outputs"].get("hidden_states") if output_hidden_states else None  # noqa: F841
+            )
+
+        # keep track of which sequences are already finished
+        # *************** Csm specific ***************
+        batch_size, cur_len = input_ids.shape[:2]
+        # ============================================
```
no longer true no

ArthurZucker replied:
?

eustlb replied:
indeed, nice catch

ArthurZucker commented on src/transformers/models/csm/modular_csm.py:
```diff
@@ -0,0 +1,1517 @@
+            backbone_last_hidden_state = outputs.hidden_states[-1][:, -1, :]
+
+            torch.compiler.cudagraph_mark_step_begin()
+            depth_decoder_outputs = self.depth_decoder.generate(
+                input_ids=first_codebook_ids,
+                backbone_last_hidden_state=backbone_last_hidden_state,
+                generation_config=depth_decoder_generation_config,
+            )
```
`self.depth_decoder`'s generation config should be `depth_decoder_generation_config` and this should not be needed no? Why is depth_decoder's generation config not the correct on?

eustlb replied:
I did all this logic because the issue is that submodels `generation_config` are not loaded properly when using `save_pretrained` and `from_pretrained`. If did a tiny fix here (see overrides for `save_pretrained` and `from_pretrained`) to ensure generation parameters with the prefix `depth_decoder_` in the model's generation config are correctly placed in the `depth_decoder`'s `generation_config`.

ArthurZucker commented on src/transformers/models/csm/modular_csm.py:
```diff
@@ -0,0 +1,1517 @@
+            next_tokens = codebook_ids
+            # ============================================
+
+            # *************** Csm specific ***************
+            # finished sentences should have their next token be a padding token
+            if has_eos_stopping_criteria:
+                next_tokens = next_tokens * unfinished_sequences.unsqueeze(-1) + pad_token_id * (
+                    1 - unfinished_sequences.unsqueeze(-1)
+                )
+            # ============================================
+
+            # *************** Csm specific ***************
+            # update generated ids, model inputs, and length for next step
+            if input_ids.ndim == 2:
+                input_ids = next_tokens[:, None, :]
+            else:
+                input_ids = torch.cat([input_ids, next_tokens[:, None, :]], dim=1)
+            # ============================================
+            if streamer is not None:
+                streamer.put(next_tokens.cpu())
+
+            # *************** Csm specific ***************
+            # for the eos stopping criteria, is it expected that the eos token is the same for each codebook !!!!
+            unfinished_sequences = unfinished_sequences & ~(
+                input_ids[:, -1, :-1] == self.config.codebook_eos_token_id
+            ).all(-1)
+            # ============================================
+            unfinished_sequences = unfinished_sequences & ~stopping_criteria(input_ids, scores)
+            this_peer_finished = unfinished_sequences.max() == 0
+            cur_len += 1
+
+            # This is needed to properly delete outputs.logits which may be very large for first iteration
+            # Otherwise a reference to outputs is kept which keeps the logits alive in the next iteration
+            del outputs
+            # *************** Csm specific ***************
+            del depth_decoder_outputs
+            # ============================================
```
some parts are csm specific should be wrapped under a single

ArthurZucker commented on src/transformers/models/csm/modular_csm.py:
```diff
@@ -0,0 +1,1517 @@
+            streamer.end()
+
+        if return_dict_in_generate:
+            if self.config.is_encoder_decoder:
```
not needed no?

eustlb replied:
indeed, kept it to ease maintenance when comparing with generate's _sample function, I can remove it though

ArthurZucker replied:
yes feel free to remove

ArthurZucker commented on src/transformers/models/csm/modular_csm.py:
```diff
@@ -0,0 +1,1517 @@
+                kept_criteria.append(criterion)
+        return kept_criteria
+
+    def _merge_input_ids_with_input_values(
```
define it above

ArthurZucker commented on src/transformers/models/csm/modular_csm.py:
```diff
@@ -0,0 +1,1517 @@
+            audio_token_mask = input_ids == audio_token_id
+
+            # retreive the number of audio tokens for each audio from the input_ids
+            change_idxs = (audio_token_mask[:, 1:] != audio_token_mask[:, :-1]).nonzero(as_tuple=True)[-1]
+            num_audio_tokens = change_idxs.diff()
+            num_audio_tokens = torch.stack([num_audio_tokens[i] for i in range(0, len(num_audio_tokens), 2)])
+
+            frames_mask = torch.arange(max_audio_frames, device=batched_audio_token_ids.device).expand(
+                len(num_audio_tokens), -1
+            ) < num_audio_tokens.unsqueeze(-1)
+            frames_mask = frames_mask.flatten()
+
+            audio_token_idxs = audio_token_mask.nonzero(as_tuple=True)
+            inputs_embeds[audio_token_idxs[0], audio_token_idxs[1]] = audio_embeds.view(-1, audio_embeds.shape[-1])[
+                frames_mask
+            ]
```
not sure this is needed, you could use token type ids no?

eustlb replied:
Not sure to see what you mean here. All this is needed to place back the audio `inputs_embeds` after correctly, since we need to:
1. inferring the codec model (in a batch padded manner, meaning one forward pass with batched stacked audio)
2. embedding those tokens
3. place back the embeds at the correct positions

ArthurZucker commented on src/transformers/models/csm/modular_csm.py:
```diff
@@ -0,0 +1,1517 @@
+        self, generation_config: Optional[GenerationConfig], use_model_defaults: Optional[bool] = None, **kwargs: Dict
+    ) -> Tuple[GenerationConfig, Dict]:
+        """
+        This method overrides [~generation.utils.GenerationMixin._prepare_generation_config].
+        It ensures that the depth decoder generation config is initialized and that passed args as depth_decoder_* are properly handled.
```
as I mentionned this is weird, initializing the depth decoder should give it its own genration config

eustlb replied:
Yep, it is now done this way. Nevertheless we still need to ensure parameters passed to the generate function specifically to control the depth decoder generation (e.g. `depth_decoder_temperature`) and correctly removed from `model_kwargs` and placed in the `depth_decoder` `generation_config`.

ArthurZucker commented on src/transformers/models/csm/modular_csm.py:
```diff
@@ -0,0 +1,1517 @@
+        output_audio: Optional[bool] = False,
+        **kwargs,
+    ) -> Union[GenerateNonBeamOutput, torch.LongTensor]:
+        r"""
```
big part of the doc is not needed

ArthurZucker commented on src/transformers/models/csm/processing_csm.py:
```diff
@@ -0,0 +1,294 @@
+    encoded_length_kwargs: Optional[Dict[str, Any]]
+
+
+class CsmProcessorKwargs(ProcessingKwargs, total=False):
```
nice!

ArthurZucker reviewed (commented):
Great work!

eustlb commented on src/transformers/models/csm/modular_csm.py:
```diff
@@ -0,0 +1,1517 @@
+
+        if inputs_embeds is None:
+            input_ids_are_first_codebook = position_ids[0, 0] == 0
+            codebook_idxs = nn.functional.relu(position_ids - 1)
```
yeah ahah this one is an easter egg

eustlb committed:
do not default to mutable

eustlb commented on src/transformers/models/csm/configuration_csm.py:
```diff
@@ -0,0 +1,506 @@
+        if self.rope_scaling is not None and "type" in self.rope_scaling:
+            self.rope_scaling["rope_type"] = self.rope_scaling["type"]
+        rope_config_validation(self)
+        self.vocab_size = vocab_size
```
Ah! Slipped through, thanks for noticing!

eustlb committed:
remove duplicates

eustlb commented on src/transformers/models/csm/modular_csm.py:
```diff
@@ -0,0 +1,1517 @@
+        self.num_codebooks = num_codebooks
+        self.weight = nn.Parameter(torch.empty(self.num_codebooks - 1, hidden_size, vocab_size))
+
+    def reset_parameters(self):
+        for i in range(self.num_codebooks - 1):
```
Yep, did that originally to override torch's `reset_parameters` default and unsure each sub-matrix corresponding to a linear layer is initialized as a linear layer but useless indeed since we have `_init_weights`

eustlb committed:
Merge branch 'main' into add-csm

eustlb committed:
remove useless reset_parameters

eustlb committed:
use GradientCheckpointingLayer

eustlb committed:
use can_return_tuple

eustlb committed:
formatting

eustlb committed:
prepend placeholder in _sample

eustlb committed:
torch compile fix

eustlb committed:
some more fixies

eustlb committed:
convert modular

eustlb committed:
Merge branch 'main' into add-csm

eustlb committed:
fix

eustlb committed:
default max_length in convert

eustlb committed:
handle depth decoder generation config correctly

eustlb commented on src/transformers/models/csm/modular_csm.py:
```diff
@@ -0,0 +1,1517 @@
+            )
+
+        if inputs_embeds is None:
+            input_ids_are_first_codebook = position_ids[0, 0] == 0
```
Changed few things here to ensure compile compatibility.

eustlb commented:
Thanks a lot @ArthurZucker for the review 🐐!
I’ve addressed most of your comments (and the rest tomorrow). I’ve introduced a few changes, which I detailed in the comments. Namely:
1. The `depth_decoder` generation parameters are stored in the model’s `generation_config.json` under the `depth_decoder` prefix, and are correctly loaded into the `depth_decoder`’s `generation_config` when using `from_pretrained` (and also properly saved).
2. Since the `depth_decoder` does not support padding or inconsistent `position_ids`, I moved away from using `position_ids` and instead rely directly on `cache_position`. However, I kept `position_ids` in the model’s forward signature to latter log users that it’s not supported for this model, since it’s a widely used parameter in the library.

ArthurZucker commented on src/transformers/models/csm/modular_csm.py:
```diff
@@ -0,0 +1,1511 @@
+            frames_mask = frames_mask.flatten()
+
+            audio_token_idxs = audio_token_mask.nonzero(as_tuple=True)
+            inputs_embeds[audio_token_idxs[0], audio_token_idxs[1]] = audio_embeds.view(-1, audio_embeds.shape[-1])[
+                frames_mask
+            ]
```
so, TLDR, you are only selecting the audio embeddings for which the number of audio tokens is smaller than the max_audio_frames. 
This begs 2 questions:
1. Why are we wasting compute on these audio frames
2. Why don't you just do something like:

```python
# Compute per-row cumulative audio token count
audio_frame_indices = audio_token_mask.cumsum(dim=1) - 1  

# Mask out tokens where frame index >= max_audio_frames
valid_audio_mask = audio_token_mask & (audio_frame_indices < max_audio_frames)

# Use that to index into inputs_embeds
inputs_embeds[valid_audio_mask] = audio_embeds.view(-1, audio_embeds.shape[-1])[
    (audio_frame_indices[valid_audio_mask]).flatten()
]
```

ArthurZucker replied:
this allows you to directly mask the entire audio embeddings that are wrong instead of re-creating the frame mask

eustlb replied:
An important thing to note is that each batch element of the inputs can have multiple audio. Breaking it down with an example: 
```python
[
	["...<|AUDIO|>...<|AUDIO|..."],
	["...<|AUDIO|>..."]
]
```

But when we infer the codec model, we infer it with all the audio in one batch, so the corresponding `batched_audio_token_ids` will be of shape `[3, max_audio_frames, num_codebooks]`.

**Why are we wasting compute on these audio frames** ?
→  Can be seen like padding. We are waisting compute on these because run batched inference (here I am faking batched inference because for Mimi it is no yet equivalent to run batch and sequential) so audio are padded and thus we

**Why don't you just do something like...**
When we retrieve the codec's batched output, idxs in `audio_token_mask` are not the same as the ones in the output because the audio inputs are stacked and padded to max length. The only correspondance we have is that we know they are stacked, right padded and how many tokens for each should be retrieved.

ArthurZucker replied:
No, I mean that the code I provided should be equivalent to the one you used but just simplified

ArthurZucker reviewed (approved):
A lot better! 
The generation loop could have it's own separate file IMO, as we did for whisper, but its not a blocker

eustlb commented on src/transformers/models/csm/modular_csm.py:
```diff
@@ -0,0 +1,1511 @@
+        self.weight = nn.Parameter(torch.empty(self.num_codebooks - 1, hidden_size, vocab_size))
+
+    def forward(self, hidden_states, cache_position):
+        if cache_position is None:
```
`cache_position` can be None during training no? We need to be able to run a forward without providing `cache_position`.

ArthurZucker replied:
`if cache_position is None:` in the model forward AFAIK always creates them

eustlb committed:
clearer formulation

eustlb committed:
handle output_loading_info

eustlb committed:
handle softmax warning

eustlb committed:
Merge branch 'main' into add-csm

eustlb committed:
Merge branch 'main' into add-csm

eustlb committed:
add doc

eustlb committed:
propagate _get_initial_cache_position changes

eustlb committed:
generation in its own module

eustlb committed:
add processor tests

eustlb commented on src/transformers/models/csm/modular_csm.py:
```diff
@@ -0,0 +1,1511 @@
+        for attr, value in depth_decoder_attrs.items():
+            setattr(self.generation_config, prefix + attr, value)
+
+        LlamaForCausalLM.save_pretrained(*args, **kwargs)
```
Totally agree here! I'll keep it like that here and generalise in a subsequent PR

eustlb commented:
I've iterated again on you're review @ArthurZucker 🤗
Mainly:
1. csm now has it's own `generation_csm.py` module
2. keeping `from_pretained` and `save_pretrained` there for the moment, I'll open another PR to generalize nested generation configs for sub models.
3. added processor tests

Just have to finish `csm.md` to make sure I enable users with all the possibilities this model offers (a special _go brrr_ section for compile and _train_ section) and we'll be good to merge!

eustlb committed:
fix compile witu cuda graphs

eustlb committed:
fix compile with cuda graphs

eustlb committed:
add csm.md

eustlb committed:
include CSM loss

eustlb committed:
Merge branch 'main' into add-csm

eustlb committed:
doc nit

eustlb committed:
Merge branch 'add-csm' of github.com:eustlb/transformers into add-csm

eustlb committed:
doc nit

eustlb committed:
doc nit

ArthurZucker commented on docs/source/en/model_doc/csm.md:
```diff
@@ -0,0 +1,406 @@
+model = CsmForConditionalGeneration.from_pretrained(model_id, device_map=device)
+
+# prepare the inputs
+text = "The past is just a story we tell ourselves."
+speaker_id = 0
+inputs = processor(f"[{speaker_id}]{text}", add_special_tokens=True).to(device)
```
```suggestion
text = "[0]The past is just a story we tell ourselves." # `[0]` for speaker id 0
inputs = processor(text, add_special_tokens=True).to(device)
```
IMO this is more explicit

ArthurZucker commented on docs/source/en/model_doc/csm.md:
```diff
@@ -0,0 +1,406 @@
+
+# infer the model
+audio_values = model.generate(**inputs, output_audio=True)
+sf.write("example_without_context.wav", audio_values[0].cpu().numpy(), 24000)
```
thinking we can have a `processor.save_audio()`

eustlb replied:
agree and done! Later to be upstreamed to `AudioProcessor`, left a TODO

ArthurZucker commented on docs/source/en/model_doc/csm.md:
```diff
@@ -0,0 +1,406 @@
+
+# 1. context
+for text, audio, speaker_id in zip(ds[:4]["text"], ds[:4]["audio"], ds[:4]["speaker_id"]):
+    conversation.append(
+        {
+            "role": f"{speaker_id}",
+            "content": [{"type": "text", "text": text}, {"type": "audio", "path": audio["array"]}],
+        }
+    )
```
can't we use a chat template for this?

ArthurZucker commented on docs/source/en/model_doc/csm.md:
```diff
@@ -0,0 +1,406 @@
+
+# compile
+model.generation_config.cache_implementation = "static"
+model.depth_decoder.generation_config.cache_implementation = "static"
+model.depth_decoder.forward = torch.compile(model.depth_decoder.forward, fullgraph=True, mode="reduce-overhead")
+model.forward = torch.compile(model.forward, fullgraph=True, mode="reduce-overhead")
```
super cool! one issue here is that you are gonna hit recompiles for the shape of `prefill` vs `decoding` at least that is the case for general LLM. Is it not the case for CSM?

eustlb replied:
True, there is indeed recompile for `prefill` vs `decoding`, happening at the first pass, that is why pad inputs to a given constant shape. Yet it is unnecessary as we can have the prefill left uncompiled. I am going to remove that, thanks.

ArthurZucker commented on docs/source/en/model_doc/csm.md:
very very nice!

ArthurZucker commented on src/transformers/models/csm/modular_csm.py:
```diff
@@ -0,0 +1,1041 @@
+from .generation_csm import CsmGenerationMixin
+
+
+if TYPE_CHECKING:
+    pass
```
```suggestion
```

ArthurZucker commented on src/transformers/models/csm/modular_csm.py:
```diff
@@ -0,0 +1,1041 @@
+            audio_token_mask = input_ids == audio_token_id
+
+            # retreive the number of audio tokens for each audio from the input_ids
+            change_idxs = (audio_token_mask[:, 1:] != audio_token_mask[:, :-1]).nonzero(as_tuple=True)[-1]
+            num_audio_tokens = change_idxs.diff()
+            num_audio_tokens = torch.stack([num_audio_tokens[i] for i in range(0, len(num_audio_tokens), 2)])
+
+            frames_mask = torch.arange(max_audio_frames, device=batched_audio_token_ids.device).expand(
+                len(num_audio_tokens), -1
+            ) < num_audio_tokens.unsqueeze(-1)
+            frames_mask = frames_mask.flatten()
+
+            inputs_embeds[audio_token_mask] = audio_embeds.view(-1, audio_embeds.shape[-1])[frames_mask]
```
I maintain that the suggestion is equivalent to what you were doing 😉 https://github.com/huggingface/transformers/pull/36719#discussion_r2067863590

ArthurZucker commented on src/transformers/models/csm/modular_csm.py:
```diff
@@ -0,0 +1,1041 @@
+        depth_decoder_outputs = None
+        if labels is not None:
+            # select first codebook as labels for the backbone model
+            backbone_labels = labels[:, :, 0] if labels is not None else None
+            backbone_loss = self.loss_function(
+                logits=backbone_logits, labels=backbone_labels, vocab_size=self.config.vocab_size, **kwargs
+            )
+
+            # for the depth decoder, we need to select the frames to train on
+            # those are frames where the label is not uniformly `ignore_index` along the codebook dimension
+            depth_decoder_labels = labels
+            mask_idxs = (depth_decoder_labels[:, :, 1:] == -100).all(dim=-1)
+            train_idxs = (~mask_idxs).nonzero(as_tuple=True)
+
+            depth_decoder_input_ids = labels[train_idxs[0], train_idxs[1], : self.config.num_codebooks - 1]
+            # adds place holder in position 0 that will be replaced by the backbone_last_hidden_state
+            depth_decoder_input_ids = nn.functional.pad(depth_decoder_input_ids, (1, 0), value=0)
+
+            backbone_last_hidden_states = backbone_hidden_states[train_idxs[0], train_idxs[1] - 1, :]
+            depth_decoder_labels = depth_decoder_labels[train_idxs[0], train_idxs[1], :]
+
+            depth_decoder_outputs = self.depth_decoder(
+                input_ids=depth_decoder_input_ids,
+                backbone_last_hidden_state=backbone_last_hidden_states,
+                use_cache=use_cache,
+                output_attentions=output_attentions,
+                output_hidden_states=output_hidden_states,
+                return_dict=True,
+                labels=depth_decoder_labels,
+            )
+
+            depth_decoder_loss = depth_decoder_outputs.loss
+            loss = backbone_loss + depth_decoder_loss
```
this should be in the `self.loss_function` with a `loss_csm.py`

eustlb replied:
not sure it really makes sense here since the logic here it just selecting the correct inputs before inferring the depth decoder to get the loss, i.e.:
1. getting the backbone loss with first call to self.loss_function
2. preparing inputs and getting the depth_decoder loss by inferring the depth decoder forward with correct labels

to have it in loss_csm.py we would have to pass the depth decoder model to `self.loss_function` which is something we want to avoid no? WDYT?

ArthurZucker reviewed (approved):
Nice! 
2 comments for me before we can merge:
- make some code simpler
- loss in a separate file!

eustlb committed:
Update docs/source/en/model_doc/csm.md

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

eustlb committed:
add save_audio to processor

eustlb committed:
Merge branch 'add-csm' of github.com:eustlb/transformers into add-csm

eustlb committed:
Update src/transformers/models/csm/modular_csm.py

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

eustlb committed:
doc update

eustlb committed:
simplify audio_codes_mask computation

eustlb committed:
doc update

eustlb committed:
Merge branch 'main' into add-csm

eustlb committed:
simplify loss computation

eustlb committed:
fix static cache test

eustlb committed:
fix

eustlb committed:
remove comment

eustlb committed:
Merge branch 'main' into add-csm

ArthurZucker commented on src/transformers/models/mimi/modeling_mimi.py:
```diff
@@ -1567,6 +1593,44 @@ def _encode_frame(
+        """
+        output_length = input_length
+
+        # encoder
```
we can get the exact layer normally

ArthurZucker replied:
we have `get_module` utiles based on naming

ArthurZucker commented on src/transformers/models/mimi/modeling_mimi.py:
```diff
@@ -216,6 +216,32 @@ def _pad1d(hidden_states: torch.Tensor, paddings: Tuple[int, int], mode: str = "
         end = padded.shape[-1] - extra_pad
         return padded[..., :end]
 
+    def _get_output_length(self, input_length: int) -> int:
```
would help you to have a tensor here no?

ArthurZucker commented on src/transformers/models/mimi/modeling_mimi.py:
```diff
@@ -1567,6 +1593,44 @@ def _encode_frame(
+        """
+        Get the mask for the audio codes from the original padding mask.
+        """
+        audio_lengths = padding_mask.sum(dim=-1).cpu().tolist()
```
would be cool to keep tensors

ArthurZucker reviewed (approved):
Let's go!

eustlb committed:
Merge branch 'main' into add-csm

eustlb committed:
simplify encoded length computation

eustlb committed:
use hf-internal-testing

eustlb committed:
Merge branch 'main' into add-csm

eustlb committed:
Merge branch 'main' into add-csm

eustlb committed:
doc update

eustlb committed:
Merge branch 'add-csm' of github.com:eustlb/transformers into add-csm

eustlb committed:
cast to float before numpy

eustlb committed:
nit

eustlb committed:
mem efficient codebook head

eustlb committed:
Merge branch 'main' into add-csm

eustlb committed:
nit

eustlb committed:
Merge branch 'add-csm' of github.com:eustlb/transformers into add-csm

eustlb committed:
Merge branch 'main' into add-csm

eustlb committed:
cat input values with cutoffs

eustlb committed:
Merge branch 'main' into add-csm

johnwick123f commented:
Not sure if I should ask here but this implementation of csm seems considerably better quality then the official implementation for some reason? Any reason, why?
